### PR TITLE
test: fix mixer_test.c:comp_register() 'uninitialized err' -Werror

### PR DESCRIPTION
**Pull request by marc-hb**
Fixes:
```
  error: 'err' may be used uninitialized in this function
                           [-Werror=maybe-uninitialized]
  if (err)
     ^
```
As I found nothing in the entire code base checking the return value of
comp_register() (the real function seems to always succeed), I also
simplify the main path and pass through the value returned by
memcpy_s().

Signed-off-by: Marc Herbert <marc.herbert@intel.com>

### CODE REVIEW

**Comment by marc-hb** [comment]
The https://sof-ci.01.org/sofpr/PR2644/build5573/devicetest failure is only the usual check-sof-logger ABI #2507

File: test/cmocka/src/audio/mixer/mixer_test.c [hunk 1]
```diff
@@ -47,12 +47,12 @@ int comp_register(struct comp_driver_info *info)
...
-	}
 
-	if (err)
-		return -EINVAL;
+	default:
+		return -ENOTSUP;
+	}
 
-	return 0;
+	return -err;
```
Comment (paulstelian97) [inline test/cmocka/src/audio/mixer/mixer_test.c]: This looks odd.

Comment (marc-hb) [reply test/cmocka/src/audio/mixer/mixer_test.c]: @paulstelian97 what does look odd, the minus sign? As I wrote in the commit message, no one seems to care about `comp_register()`'s return value yet so I can change it to `return err` if you wish.

Comment (paulstelian97) [reply test/cmocka/src/audio/mixer/mixer_test.c]: You should return err so a negative return value is given on error. We shouldn't count on the "nobody cares yet" thing for a rather critical function like this one.

Comment (paulstelian97) [reply test/cmocka/src/audio/mixer/mixer_test.c]: I mean if you do "return err" you would in fact not change observable behavior as memcpy_s would only return 0 or -EINVAL.

Comment (marc-hb) [reply test/cmocka/src/audio/mixer/mixer_test.c]: I didn't look at `sof/src/arch/xtensa/include/arch/string.h` and I didn't expect `memcpy_s()` to return negative errors. AFAICT the documentation from Microsoft (who lobbied for these optional functions, not available on Linux) refers to positive failure codes:
https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/memcpy-s-wmemcpy-s Other documentation I found also seemed to refer to positive values too.
The standard only says "non-zero". Negative errors are useful to leave room for multiple positive success values but `memcpy_s()` has only one form of success: zero.

tl;dr: I suspect the negative values in `sof/src/arch/xtensa/include/arch/string.h` are a mistake but I don't mind so unless someone else says something I'll go and change it to `return err;`

PS: not sure why `comp_register()` should return negative failures either but I digress :-)

**Comment by marc-hb** [comment]
Closing and re-opening to try to run "Internal Intel CI System/merge/build" that never ran.

**Comment by marc-hb** [comment]
device-test errors:
- the usual #2507 ABI line
- 3 failures on APL_UP2_NOCODEC (which seems the less stable of them all)
     - 2  `Unable to install hw params`, I've seen those before
     - 1 `dsp is not suspended after 15s, end test` in `check-kmod-load-unload.sh`

None can be relevant.

File: test/cmocka/src/audio/mixer/mixer_test.c (hunk 1 above)
Comment (paulstelian97) [reply test/cmocka/src/audio/mixer/mixer_test.c]: In the Linux kernel and SOF firmware negative values are almost always considered errors, and positive and 0 values as valid results. Microsoft sure uses another convention for reporting errors by returning a separate error code besides the return value but Linux and SOF don't have this possibility.

The Linux userspace does something like the following in system calls:

```c
long syscall(long number, ...) {
    long result;
    /* Get the arguments in arg1, arg2, ... */
    asm volatile("int 0x80\n\t"; "=a"(number), "=b"(arg1), ...; "a"(result); /* clobbers */);
    if (result < 0) {
        errno = -result; /* so errno is a positive value */
        result = -1;
    }
}
```

Comment (marc-hb) [reply test/cmocka/src/audio/mixer/mixer_test.c]: Thanks @paulstelian97.   I don't think the Linux _kernel_ is a good reference. Please correct me but the Linux kernel rarely ever copies a function from somewhere else keeping the exact same name and prototype and changing just the return values. That's error-prone.

Negative errors everywhere is simple and great, however and unlike SOF  the Linux kernel doesn't try to mimic existing userspace APIs.

> Microsoft sure uses another convention for reporting errors by returning a separate error code besides the return value but Linux and SOF don't have this possibility

I guess you are referring to the `errno`  **variable** which is a (not Microsoft specific) solution to an unrelated issue that doesn't apply to functions like `memcpy_s()` and others since they don't have anything to return on success.

Is there SOF documentation on these topics?

Comment (paulstelian97) [reply test/cmocka/src/audio/mixer/mixer_test.c]: SOF doesn't use errno and doesn't actually try that hard to emulate userspace APIs. In fact checkpatch.pl that runs on every single pull request is precisely the one from Linux, with very minor changes to make it compatible with SOF.

I meant GetLastError(), not the errno variable.

Also, SOF is called _by_ the kernel so it kinda has to integrate well into it. If it returns -ENOMEM, the kernel will see -ENOMEM and will forward it to the system call interface.

I'm not aware of any documentation for this specifically. It's just that it's been intended from the get go to emulate coding style, overall design etc from the Linux kernel. There are more similarities than differences in this sense.

Comment (marc-hb) [reply test/cmocka/src/audio/mixer/mixer_test.c]: > SOF doesn't actually try that hard to emulate userspace APIs.

SOF has #includes with familiar names and a dependency on an unspecified number of [newlib](http://www.sourceware.org/newlib/) headers. While I probably overestimated the importance of these, I don't think the Linux kernel does something of that magnitude. In any case it'd be amazed if the kernel copies an existing  function prototype exactly while changing only the sign of return values, something the compiler has no way to check.

> In fact checkpatch.pl that runs on every single pull request is precisely the one from Linux, with very minor changes to make it compatible with SOF.

I just added one of these "minor changes" yet I had no idea checkpatch.pl was performing anything related to this APIs and return values question. Example?

> Also, SOF is called by the kernel so it kinda has to integrate well into it. If it returns -ENOMEM, the kernel will see -ENOMEM and will forward it to the system call interface.

Interesting, I didn't expect the return value of low-level functions like `memcpy_s` to be ever sent "as is" to the kernel, do you have examples?

To be clear I never suggested SOF should avoid negative return values. I'm merely suggesting it should avoid existing names like `memcpy_s` unless it copies these functions exactly. `sof_memcpy_s` with negative return values would have been perfect.

> There are more similarities than differences in this sense.

I do see similarities with the Linux kernel (and userspace) of course but none relevant to this discussion about return values.

Comment (paulstelian97) [reply test/cmocka/src/audio/mixer/mixer_test.c]: memcpy_s isn't forced to trigger an immediate crash, it could also integrate and return an error code. The convention within SOF is to return it as a negative value (so that calling code could do `if (ret < 0) goto err;` or similar).

SOF doesn't communicate with the userspace really but with the kernel space.

Additionally, according to the standard memcpy_s just returns a nonzero value. However the rest of the functions return negative values for errors and 0 or positive for success. It would be much more of a source of bugs to have one function that works differently from the rest of the codebase.

You should expect with low level code such as SOF and any OS kernel that even "standard" functions may behave differently from the same function in another project. memcpy_s behaves in a compatible way (-EINVAL is still nonzero and thus still standards compliant).
